plugins/out_s3/s3.c

```diff
@@ -1243,8 +1303,27 @@ static int put_all_chunks(struct flb_s3 *ctx)
                                 chunk->create_time, buffer, buffer_size);
             flb_free(buffer);
             if (ret < 0) {
-                s3_store_file_unlock(chunk);
                 chunk->failures += 1;
+                if (ctx->has_old_buffers == FLB_FALSE) {
+                    if (chunk->failures > ctx->ins->retry_limit) {
+                        s3_retry_warn(ctx, (char *) fsf->meta_buf, NULL,
+                                      chunk->create_time, FLB_FALSE);
+                        s3_store_file_delete(ctx, chunk);
+                    }
+                    else {
+                        s3_retry_warn(ctx, (char *) fsf->meta_buf, NULL,
+                                      chunk->create_time, FLB_TRUE);
+                        s3_store_file_unlock(chunk);
+                    }
+                }
+                else {
+                    if (chunk->failures > ctx->ins->retry_limit) {
+                        flb_plg_error(ctx->ins, "failed to flush chunk tag=%s, "
+                                      "create_time=%s, chunk will remain buffered "
+                                      "on filesystem", (char *) fsf->meta_buf,
+                                      chunk->create_time);
+                    }
+                }
                 return -1;
```

Comment:
This is shutdown case? If so, we want to print this warning for all failures. Not just when the failures is greater than retry_limit. Do you understand why?

Reply:
Since the logic is inverted I think this is actually the startup case. In which case, the `failures` will be 0 on startup and so we need to print the "will retry" version of the message.